storage/scylladb_permanode.c

```diff
@@ -0,0 +1,614 @@
+    return SC_TA_NULL;
+  }
+
+  if (prepare_query(session, query, &prepared) != CASS_OK) {
+    ta_log_error("Fail to prepare query : %s\n", query);
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+  statement = ret_insert_transaction_statement(prepared, hash, trytes);
+  if (execute_statement(session, statement) != CASS_OK) {
+    ta_log_error("Fail to insert transactions into transaction table\n");
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+  cass_prepared_free(prepared);
+  return ret;
+}
+
+static status_t insert_approvee(CassSession* session, const tryte_t* hash, const tryte_t* approvee) {
+  status_t ret = SC_OK;
+  const CassPrepared* prepared = NULL;
+  CassStatement* statement = NULL;
+  CassCollection* collection = NULL;
+
+  if (hash == NULL || approvee == NULL) {
+    ta_log_error("NULL pointer to ScyllaDB transaction object\n");
+    return SC_TA_NULL;
+  }
+
+  char* query = "UPDATE transaction SET approvees = approvees + ? WHERE hash = ?";
+  if (prepare_query(session, query, &prepared) != CASS_OK) {
+    ta_log_error("Fail to prepare query : %s\n", query);
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+  statement = cass_prepared_bind(prepared);
+  collection = cass_collection_new(CASS_COLLECTION_TYPE_SET, 1);
+  cass_collection_append_bytes(collection, (cass_byte_t*)approvee, NUM_TRYTES_HASH);
+  cass_statement_bind_collection(statement, 0, collection);
+  cass_collection_free(collection);
+  cass_statement_bind_bytes(statement, 1, (cass_byte_t*)hash, NUM_TRYTES_HASH);
+  if (execute_statement(session, statement) != CASS_OK) {
+    ta_log_error("Fail to insert approvee\n");
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+  cass_prepared_free(prepared);
+exit:
+
+  return ret;
+}
+
+static status_t push_collection_into_queue(CassSession* session, CassStatement* statement,
+                                           hash243_queue_t* const res_queue, const char* column_name) {
+  status_t ret = SC_CCLIENT_NOT_FOUND;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+
+  result = cass_future_get_result(future);
+  iterator = cass_iterator_from_result(result);
+  if (cass_iterator_next(iterator)) {
+    const CassValue* value = NULL;
+    const CassRow* row = cass_iterator_get_row(iterator);
+    CassIterator* items_iterator = NULL;
+    value = cass_row_get_column_by_name(row, column_name);
+
+    items_iterator = cass_iterator_from_collection(value);
+    while (cass_iterator_next(items_iterator)) {
+      size_t len;
+      const cass_byte_t* buf;
+      cass_value_get_bytes(cass_iterator_get_value(items_iterator), &buf, &len);
+      if (hash243_queue_push(res_queue, (flex_trit_t*)buf) != RC_OK) {
```

Comment:
Does the data saved in cassandra are in `flex_trit_t`?

Reply:
It is `cass_byte_t` which is actually `uint8_t`. Same as `flex_trit_t`.